### PR TITLE
gh-127065: Make methodcaller thread-safe and re-entrant (v2)

### PR DESCRIPTION
**Pull request by eendebakpt**
See #127245 for details.

* Issue: gh-127065

### CODE REVIEW

File: Modules/_operator.c [hunk 1]
```diff
@@ -1745,8 +1748,10 @@ static int
 methodcaller_traverse(methodcallerobject *mc, visitproc visit, void *arg)
 {
     Py_VISIT(mc->name);
-    Py_VISIT(mc->xargs);
+    Py_VISIT(mc->args);
     Py_VISIT(mc->kwds);
+    Py_VISIT(mc->vectorcall_args);
+    Py_VISIT(mc->vectorcall_kwnames);
```
Comment (eendebakpt) [inline Modules/_operator.c:1750]: Note `mc->vectorcall_kwnames` is a tuple of  strings, but the strings can be subclasses of `str`, so we need to visit this.

File: Modules/_operator.c [hunk 2]
```diff
@@ -1595,78 +1595,79 @@ static PyType_Spec attrgetter_type_spec = {
...
+static int
+_methodcaller_initialize_vectorcall(methodcallerobject* mc)
+{
+    PyObject* args = mc->args;
+    PyObject* kwds = mc->kwds;
+
+    if (kwds && PyDict_Size(kwds)) {
+        PyObject *values = PyDict_Values(kwds);
+        if (!values) {
+            PyErr_NoMemory();
```
Comment (colesbury) [inline Modules/_operator.c]: I don't think you should call `PyErr_NoMemory()` here (or in the cases below). `PyDict_Values()` should set the appropriate exception on error.

Most Python C APIs set the appropriate error on out-of-memory with the exception of some "lower level" APIs like `PyObject_Malloc()` and `PyMem_Malloc()`.

File: Modules/_operator.c [hunk 3]
```diff
@@ -1595,78 +1595,79 @@ static PyType_Spec attrgetter_type_spec = {
...
+                return -1;
+            }
+        }
+        else {
+            mc->vectorcall_args = values_tuple;
+        }
+        mc->vectorcall_kwnames = PySequence_Tuple(kwds);
+        if (!mc->vectorcall_kwnames) {
+                PyErr_NoMemory();
+                return -1;
```
Comment (colesbury) [inline Modules/_operator.c]: formatting: indentation needs to be fixed

File: Modules/_operator.c [hunk 4]
```diff
@@ -1595,78 +1595,79 @@ static PyType_Spec attrgetter_type_spec = {
...
+        PyObject *values_tuple = PySequence_Tuple(values);
+        Py_DECREF(values);
+        if (!values_tuple) {
+            PyErr_NoMemory();
+            return -1;
+        }
+        if (PyTuple_GET_SIZE(args)) {
+            mc->vectorcall_args = PySequence_Concat(args, values_tuple);
+            Py_DECREF(values_tuple);
+            if (mc->vectorcall_args == 0) {
```
Comment (colesbury) [inline Modules/_operator.c]: `mc->vectorcall_args` is a pointer, so I think either:

`if (!mc->vectorcall_args)` like you write above or
`if (mc->vectorcall_args == NULL)`

is a little clearer.

File: Modules/_operator.c [hunk 5]
```diff
@@ -1595,78 +1595,79 @@ static PyType_Spec attrgetter_type_spec = {
...
+                return -1;
+        }
+    }
+    else {
+        mc->vectorcall_args = Py_NewRef(args);
+        mc->vectorcall_kwnames = NULL;
+    }
+
+    mc->vectorcall = (vectorcallfunc)methodcaller_vectorcall;
+    return 1;
```
Comment (colesbury) [inline Modules/_operator.c]: Most Python APIs that return status codes use `0` on success (and `-1` on error) instead of `1` on success.

**Review by colesbury — COMMENTED** [review]
Looks good. A few small comments below

**Review by colesbury — APPROVED** [review]
LGTM. Thanks for fixing this!

**Pull request merged**